### PR TITLE
[enhancement] Added ability to modify setting order.

### PR DESCRIPTION
**Pull request by scorbett123**
This enhanced machine learning allows for the filtering of module lists. Either through custom priority, the amount of clicks, or just alphabetically.

Unfortunately, I had to change the names of 2 other settings due to adding a setting to the base Abstract Module that was named the same.

### CODE REVIEW

**Comment by scorbett123** [comment]
Part of #1648

File: src/main/kotlin/org/kamiblue/client/module/AbstractModule.kt [hunk 1]
```diff
@@ -52,6 +56,7 @@ abstract class AbstractModule(
     }
 
     fun toggle() {
+        clicks.value++
```
Comment (l1ving) [inline src/main/kotlin/org/kamiblue/client/module/AbstractModule.kt:59]: I think this should only be on `enable()` as some modules disable themselves

File: src/main/kotlin/org/kamiblue/client/gui/clickgui/KamiClickGui.kt [hunk 2]
```diff
@@ -89,4 +89,23 @@ object KamiClickGui : AbstractKamiGui<ModuleSettingWindow, AbstractModule>() {
...
+    fun reorderModules() {
+        val allButtons = ModuleManager.modules
+            .groupBy { it.category.displayName }
+            .mapValues { (_, modules) -> modules.map { ModuleButton(it) } }
+
+        moduleWindows.forEach { window ->
+            window.children.clear()
+            allButtons[window.originalName]?.let { window.children.addAll(it.customSort()) }
+        }
+    }
```
Comment (l1ving) [inline src/main/kotlin/org/kamiblue/client/gui/clickgui/KamiClickGui.kt:102]: You shouldn't be duplicating code, make onDisplayed just call this function

File: src/main/kotlin/org/kamiblue/client/module/modules/combat/AutoOffhand.kt [hunk 3]
```diff
@@ -63,7 +63,7 @@ internal object AutoOffhand : Module(
     private val checkCACrystal by setting("Check Crystal Aura C", false, { type == Type.CRYSTAL && offhandCrystal })
 
     // General
-    private val priority by setting("Priority", Priority.HOTBAR)
+    private val priorityInv by setting("Priority", Priority.HOTBAR)
```
Comment (l1ving) [inline src/main/kotlin/org/kamiblue/client/module/modules/combat/AutoOffhand.kt:66]: I'd much rather the abstract inherited `priority` setting be called something more verbose then making people rename their modules

**Review by l1ving — CHANGES_REQUESTED** [review]

**Comment by l1ving** [comment]
I'm not quite sure what the how this relates to #1648 though

**Comment by scorbett123** [comment]
> I'm not quite sure what the how this relates to #1648 though

Sort of fits in with Frequently used.

**Comment by l1ving** [comment]
Ah, I see. My thinking is this might conflict with `User custom module categories`, depending on how that is implemented.

My intention behind the idea is the user would be able to open the "category editing" and from there drag modules between any category in any order they want, and make their own categories. Preferably with a way to just hide modules from editing and not display them ever.

**Comment by scorbett123** [comment]
> Ah, I see. My thinking is this might conflict with `User custom module categories`, depending on how that is implemented.
> 
> My intention behind the idea is the user would be able to open the "category editing" and from there drag modules between any category in any order they want, and make their own categories. Preferably with a way to just hide modules from editing and not display them ever.

That would probably super seed the custom priorities, however I don't think that this will conflict, the order is calculated based on module list, in the GUI, so a custom category wouldn't cause any issues (I think). The dragging of modules around is inherently better, however that could 1. build on this, 2. If they want to this can be removed.

File: src/main/kotlin/org/kamiblue/client/gui/clickgui/KamiClickGui.kt [hunk 4]
```diff
@@ -89,4 +82,23 @@ object KamiClickGui : AbstractKamiGui<ModuleSettingWindow, AbstractModule>() {
...
+
+    private fun List<ModuleButton>.customSort(): List<ModuleButton> {
+        return when (ClickGUI.sortBy.value) {
+            ClickGUI.SortByOptions.CUSTOM -> this.sortedBy { -it.module.priorityForGui.value }
+            ClickGUI.SortByOptions.FREQUENCY -> this.sortedBy { -it.module.clicks.value }
```
Comment (Xiaro) [inline src/main/kotlin/org/kamiblue/client/gui/clickgui/KamiClickGui.kt:100]: Use `.sortedByDecending` if you want it in reversed ordering

File: src/main/kotlin/org/kamiblue/client/module/AbstractModule.kt [hunk 5]
```diff
@@ -33,9 +35,11 @@ abstract class AbstractModule(
     private val enabled = BooleanSetting("Enabled", false, { false }).also(::addSetting)
     private val visible = BooleanSetting("Visible", showOnArray).also(::addSetting)
     private val default = BooleanSetting("Default", false, { settingList.isNotEmpty() }).also(::addSetting)
+    val priorityForGui = IntegerSetting("Priority In GUI", 0, IntRange(0, 1000), 1, { ClickGUI.sortBy.value == ClickGUI.SortByOptions.CUSTOM }).also(::addSetting)
+    val clicks = IntegerSetting("Clicks", 0, IntRange(0, Int.MAX_VALUE), 1, { false }).also(::addSetting) // Not nice, however easiest way to save it.
```
Comment (Xiaro) [inline src/main/kotlin/org/kamiblue/client/module/AbstractModule.kt:39]: Use `<min>..<max>` instead if you want to construct an IntRange
```suggestion
    val priorityForGui = IntegerSetting("Priority In GUI", 0, 0..1000, 1, { ClickGUI.sortBy.value == ClickGUI.SortByOptions.CUSTOM }).also(::addSetting)
    val clicks = IntegerSetting("Clicks", 0, 0..Int.MAX_VALUE, 1, { false }).also(::addSetting) // Not nice, however easiest way to save it.
```

File: src/main/kotlin/org/kamiblue/client/module/AbstractModule.kt [hunk 6]
```diff
@@ -33,9 +35,11 @@ abstract class AbstractModule(
     private val enabled = BooleanSetting("Enabled", false, { false }).also(::addSetting)
     private val visible = BooleanSetting("Visible", showOnArray).also(::addSetting)
     private val default = BooleanSetting("Default", false, { settingList.isNotEmpty() }).also(::addSetting)
+    val priorityForGui = IntegerSetting("Priority In GUI", 0, IntRange(0, 1000), 1, { ClickGUI.sortBy.value == ClickGUI.SortByOptions.CUSTOM }).also(::addSetting)
```
Comment (Xiaro) [inline src/main/kotlin/org/kamiblue/client/module/AbstractModule.kt:38]: Would be better to not have small step value for setting with big range. Set the step to 50, and fineStep 1 I think.

**Review by Xiaro — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/kamiblue/client/gui/clickgui/KamiClickGui.kt (hunk 4 above)
Comment (scorbett123) [reply src/main/kotlin/org/kamiblue/client/gui/clickgui/KamiClickGui.kt:100]: Thanks

**Review by l1ving — APPROVED** [review]

**Comment by l1ving** [comment]
I cannot get the frequency mode to work, I tried toggling a couple modules a couple dozen times via the GUI, the other two modes work.

**Comment by scorbett123** [comment]
> I cannot get the frequency mode to work, I tried toggling a couple modules a couple dozen times via the GUI, the other two modes work.

Did you restart the GUI after toggling them?

**Comment by l1ving** [comment]
How?

**Comment by Xiaro** [comment]
> How?

reopen the gui

**Comment by l1ving** [comment]
Yes, I did.

**Comment by scorbett123** [comment]
Ok, will look at this tomorrow.

**Comment by scorbett123** [comment]
@l1ving Sorry my mistake, the gui calls on toggle, however I moved it to on enable. It should work now.

File: src/main/kotlin/org/kamiblue/client/module/AbstractModule.kt [hunk 7]
```diff
@@ -53,9 +57,11 @@ abstract class AbstractModule(
 
     fun toggle() {
         enabled.value = !enabled.value
+        if (enabled.value) clicks.value++
     }
 
     fun enable() {
+        clicks.value++
         enabled.value = true
     }
```
Comment (l1ving) [inline src/main/kotlin/org/kamiblue/client/module/AbstractModule.kt:66]: Why are you increasing clicks twice per enable?

Comment (scorbett123) [reply src/main/kotlin/org/kamiblue/client/module/AbstractModule.kt:66]: What do you mean, the toggle never calls enable and the enable never calls toggle. This was what was causing it not to work earlier. It needs to be incremented if either one is run.

Comment (l1ving) [reply src/main/kotlin/org/kamiblue/client/module/AbstractModule.kt:66]: Ohh oopsies

**Review by l1ving — APPROVED** [review]

**Comment by l1ving** [comment]
Thanks!

**Comment by scorbett123** [comment]
Thanks.

**Pull request merged**